encoding/ccf/encode.go

```diff
@@ -0,0 +1,1766 @@
+//
+//	nil
+//	/ simple-type-value
+//	/ optional-type-value
+//	/ varsized-array-type-value
+//	/ constsized-array-type-value
+//	/ dict-type-value
+//	/ struct-type-value
+//	/ resource-type-value
+//	/ contract-type-value
+//	/ event-type-value
+//	/ enum-type-value
+//	/ struct-interface-type-value
+//	/ resource-interface-type-value
+//	/ contract-interface-type-value
+//	/ function-type-value
+//	/ reference-type-value
+//	/ restricted-type-value
+//	/ capability-type-value
+//	/ type-value-ref
+//
+// TypeValue is used differently from inline type or type definition.
+// Inline type and type definition are used to decode exported value,
+// while TypeValue is exported value, which type is cadence.MetaType.
+// Thus, TypeValue can encode more information than type or type definition.
+func (e *Encoder) encodeTypeValue(typ cadence.Type, visited ccfTypeIDByCadenceType) error {
+	switch typ.(type) {
+	case cadence.CompositeType, cadence.InterfaceType:
+		cadenceTypeID := typ.ID()
+		if ccfID, ok := visited[cadenceTypeID]; ok {
+			// Encode visited composite/interface type value
+			// using CCF type id for compactness.
+			return e.encodeTypeValueRef(ccfID)
+		}
+		// type value ccf type id is index of visited composite/interface
+		// type value depth first.
+		visited[cadenceTypeID] = ccfTypeID(len(visited))
+	}
+
+	simpleTypeID, ok := simpleTypeIDByType(typ)
+	if ok {
+		return e.encodeSimpleTypeValue(simpleTypeID)
+	}
+
+	switch typ := typ.(type) {
+	case *cadence.OptionalType:
+		return e.encodeOptionalTypeValue(typ, visited)
+
+	case *cadence.VariableSizedArrayType:
+		return e.encodeVarSizedArrayTypeValue(typ, visited)
+
+	case *cadence.ConstantSizedArrayType:
+		return e.encodeConstantSizedArrayTypeValue(typ, visited)
+
+	case *cadence.DictionaryType:
+		return e.encodeDictTypeValue(typ, visited)
+
+	case *cadence.StructType:
+		return e.encodeStructTypeValue(typ, visited)
+
+	case *cadence.ResourceType:
+		return e.encodeResourceTypeValue(typ, visited)
+
+	case *cadence.EventType:
+		return e.encodeEventTypeValue(typ, visited)
+
+	case *cadence.ContractType:
+		return e.encodeContractTypeValue(typ, visited)
+
+	case *cadence.StructInterfaceType:
+		return e.encodeStructInterfaceTypeValue(typ, visited)
+
+	case *cadence.ResourceInterfaceType:
+		return e.encodeResourceInterfaceTypeValue(typ, visited)
+
+	case *cadence.ContractInterfaceType:
+		return e.encodeContractInterfaceTypeValue(typ, visited)
+
+	case *cadence.FunctionType:
+		return e.encodeFunctionTypeValue(typ, visited)
+
+	case *cadence.ReferenceType:
+		return e.encodeReferenceTypeValue(typ, visited)
+
+	case *cadence.RestrictedType:
+		return e.encodeRestrictedTypeValue(typ, visited)
+
+	case *cadence.CapabilityType:
+		return e.encodeCapabilityTypeValue(typ, visited)
+
+	case *cadence.EnumType:
+		return e.encodeEnumTypeValue(typ, visited)
+
+	case nil:
+		return e.encodeNilTypeValue()
+
+	default:
+		panic(fmt.Errorf("unsupported type value %s (%T)", typ.ID(), typ))
```

Comment:
```suggestion
		panic(errors.NewUnexpectedError("unsupported type value %s (%T)", typ.ID(), typ))
```

Reply:
Done in commit 4aa5f6b302b01140ec109f5c4eb5745137c1a2b4.